// @cubos/kube-templates/tests/postgres/minorUpgrade.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var kubernetes_1 = require("../../src/kubernetes");
var postgres_1 = require("../../src/postgres");
var helpers_1 = require("../helpers");
var helpers_2 = require("./helpers");
describe("postgres", function () {
    var namespace = "test-" + helpers_1.randomSuffix();
    beforeAll(function () {
        helpers_1.apply(new kubernetes_1.Namespace({
            name: namespace,
        }));
    });
    afterAll(function () {
        helpers_1.deleteObject("namespace", namespace);
    });
    test("Upgrade database from 11.0 to 11.7 without data loss", function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    helpers_1.apply(new postgres_1.Postgres({
                        name: "postgres",
                        namespace: namespace,
                    }, {
                        cpu: {
                            limit: 1,
                            request: 0,
                        },
                        memory: "64Mi",
                        version: "11.0",
                    }));
                    helpers_1.waitPodReady(namespace, "postgres-0");
                    _a = expect;
                    return [4 /*yield*/, helpers_2.queryPostgres(namespace, "postgres-0", "SELECT version()")];
                case 1:
                    _a.apply(void 0, [(_d.sent())[0]
                            .version]).toMatch(/PostgreSQL 11.0/);
                    return [4 /*yield*/, helpers_2.queryPostgres(namespace, "postgres-0", "CREATE TABLE foo (value INT)")];
                case 2:
                    _d.sent();
                    return [4 /*yield*/, helpers_2.queryPostgres(namespace, "postgres-0", "INSERT INTO foo VALUES (354687)")];
                case 3:
                    _d.sent();
                    helpers_1.apply(new postgres_1.Postgres({
                        name: "postgres",
                        namespace: namespace,
                    }, {
                        cpu: {
                            limit: 1,
                            request: 0,
                        },
                        memory: "64Mi",
                        version: "11.7",
                    }));
                    helpers_1.waitPodReady(namespace, "postgres-0");
                    _b = expect;
                    return [4 /*yield*/, helpers_2.queryPostgres(namespace, "postgres-0", "SELECT version()")];
                case 4:
                    _b.apply(void 0, [(_d.sent())[0]
                            .version]).toMatch(/PostgreSQL 11.7/);
                    _c = expect;
                    return [4 /*yield*/, helpers_2.queryPostgres(namespace, "postgres-0", "SELECT * FROM foo")];
                case 5:
                    _c.apply(void 0, [_d.sent()]).toEqual([{ value: 354687 }]);
                    return [2 /*return*/];
            }
        });
    }); });
});
